core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala

```diff
@@ -0,0 +1,207 @@
+ * a daemon to launch drivers as Mesos frameworks upon request.
+ */
+private[mesos] class MesosClusterDispatcher(
+    args: ClusterDispatcherArguments,
+    conf: SparkConf)
+  extends Logging {
+
+  private def publicAddress(conf: SparkConf, host: String): String = {
+    val envVar = conf.getenv("SPARK_PUBLIC_DNS")
+    if (envVar != null) envVar else host
+  }
+
+  private val recoveryMode = conf.get("spark.deploy.recoveryMode", "NONE").toUpperCase()
+  logInfo("Recovery mode in Mesos dispatcher set to: " + recoveryMode)
+
+  private val engineFactory = recoveryMode match {
+    case "NONE" => new BlackHoleMesosClusterPersistenceEngineFactory
+    case "ZOOKEEPER" => new ZookeeperMesosClusterPersistenceEngineFactory(conf)
+  }
+
+  private val scheduler = new MesosClusterSchedulerDriver(engineFactory, conf)
+
+  private val server = new MesosRestServer(args.host, args.port, conf, scheduler)
+  private val webUi = new MesosClusterUI(
+    new SecurityManager(conf),
+    args.webUiPort,
+    conf,
+    publicAddress(conf, args.host),
+    scheduler)
+
+  private val shutdownLatch = new CountDownLatch(1)
+  private val sparkHome = new File(Option(conf.getenv("SPARK_HOME")).getOrElse("."))
+
+  def start(): Unit = {
+    webUi.bind()
+    scheduler.frameworkUrl = webUi.activeWebUiUrl
+    scheduler.start()
+    server.start()
+  }
+
+  def awaitShutdown(): Unit = {
+    shutdownLatch.await()
+  }
+
+  def stop(): Unit = {
+    webUi.stop()
+    server.stop()
+    scheduler.stop()
+    shutdownLatch.countDown()
+  }
+}
+
+private[mesos] object MesosClusterDispatcher extends spark.Logging {
+  def main(args: Array[String]) {
+    SignalLogger.register(log)
+
+    val conf = new SparkConf
+    val dispatcherArgs = new ClusterDispatcherArguments(args, conf)
+
+    conf.setMaster(dispatcherArgs.masterUrl)
+    conf.setAppName("Mesos Cluster Dispatcher")
+
+    dispatcherArgs.zookeeperUrl.foreach { z =>
+      conf.set("spark.deploy.recoveryMode", "ZOOKEEPER")
+      conf.set("spark.deploy.zookeeper.url", z)
+    }
+
+    val dispatcher = new MesosClusterDispatcher(
+      dispatcherArgs,
+      conf)
+
+    dispatcher.start()
+
+    val shutdownHook = new Thread() {
+      override def run() {
+        logInfo("Shutdown hook is shutting down dispatcher")
+        dispatcher.stop()
+        dispatcher.awaitShutdown()
+      }
+    }
+
+    Runtime.getRuntime.addShutdownHook(shutdownHook)
+
+    dispatcher.awaitShutdown()
+  }
+
+  private class ClusterDispatcherArguments(args: Array[String], conf: SparkConf) {
+    var host = Utils.localHostName()
+    var port = 7077
+    var webUiPort = 8081
+    var masterUrl: String = _
+    var zookeeperUrl: Option[String] = None
+    var propertiesFile: String = _
+
+    parse(args.toList)
+
+    propertiesFile = Utils.loadDefaultSparkProperties(conf, propertiesFile)
+
+    def parse(args: List[String]): Unit = args match {
```

Comment:
Would be good to add `@trailrec`, making sure this method is indeed tail-recursive.